File: 0052-n-queens-ii/0052-n-queens-ii.java
class Solution {
    int cnt;
    public int totalNQueens(int n) {
        if(n <= 0 || n == 2 || n == 3){
            return 0;
        }

        cnt = 0;
        int[][] board = new int[n][n];
        helper(board,0,n);

        return cnt;
        
    }
    private void helper(int[][] board, int row, int n){
        if(row == n){
            cnt++;
            return;
        }

        for(int col = 0; col < n; col++){
            if(isSafe(board,row,col,n)){
                board[row][col] = 1;
                helper(board,row+1,n);
                board[row][col] = 0;
            }
        }
    }
    private boolean isSafe(int[][] board, int row, int col, int n){
        for(int i = row; i >= 0; i--){
            if(board[i][col] == 1) return false;
        }

        int i = row; int j = col;

        while(i >= 0 && j >= 0){
            if(board[i][j] == 1) return false;
            i--;
            j--;
        }

        i = row; j = col;

        while(i >= 0 && j < n){
            if(board[i][j] == 1) return false;
            i--;
            j++;
        }

        return true;
    }
}

File: 0052-n-queens-ii/0052-n-queens-ii.py
class Solution(object):
    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        self.cnt = 0
        board = [["."] * n for _ in range(n)]

        def is_safe(row, col):
            for i in range(row):
                if board[i][col] == "Q":
                    return False
            
            i, j = row, col
            while i >= 0 and j >= 0:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j -= 1
            
            i, j = row, col
            while i >= 0 and j < n:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j += 1
            
            return True
        
        def backtrack(row):
            if row == n:
                self.cnt += 1
                return
            
            for col in range(n):
                if is_safe(row, col):
                    board[row][col] = "Q"
                    backtrack(row + 1)
                    board[row][col] = "."  
        
        backtrack(0)
        return self.cnt

        